---
title: "Untitled"
author: "Arun Kumar"
date: "2025-05-20"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
Sys.setenv(XML_CONFIG = "/usr/bin/xml2-config")
install.packages("XML", type = "source")

```

```{r, echo=FALSE}
library(TCGAbiolinks)
library(tidyverse)
library(maftools)
library(pheatmap)
library(SummarizedExperiment)
library(edgeR)
library(limma)

```

```{r}
# get a list of projects
gdcprojects <- getGDCprojects()
getProjectSummary('TCGA-LIHC')
```

```{r}
# building a query
query_TCGA <- GDCquery(project = 'TCGA-LIHC',
         data.category = 'Transcriptome Profiling')
output_query_TCGA <- getResults(query_TCGA)
```


```{r}
query_TCGA = GDCquery(
  project = "TCGA-LIHC",
  data.category = "Transcriptome Profiling", 
  experimental.strategy = "RNA-Seq",
  workflow.type = "STAR - Counts",
  access = 'open',
                       sample.type = c("Primary Tumor","Solid Tissue Normal"))
```


```{r}
## download whole lihc data for tumore and control
GDCdownload(
  query = query_TCGA,
            directory = "/Volumes/NGS/tcga_geneexpression/Gene-Expression-Analysis/GDCdata",
  files.per.chunk = 10)

```

```{r}
TCGA_data <- GDCprepare(query_TCGA)

TCGA_data

```

```{r}
saveRDS(object = TCGA_data,
        file = "lihc_data.RDS",
        compress = FALSE)
```


```{r}
lich_data <- readRDS("lihc_data.RDS")
lich_data
```
######################################################################### 

## count to DEGs

```{r}
limma_pipeline1 = function(
  tcga_data,
  condition_variable,
  reference_group=NULL){

  design_factor = colData(tcga_data)[, condition_variable, drop=T]

  group = factor(design_factor)
  if(!is.null(reference_group)){group = relevel(group, ref=reference_group)}

  design = model.matrix(~ group)

  dge = DGEList(counts=assay(tcga_data),
                 samples=colData(tcga_data),
                 genes=as.data.frame(rowData(tcga_data)))

  # filtering
  keep = filterByExpr(dge,design)
  dge = dge[keep,,keep.lib.sizes=FALSE]
  rm(keep)

  # Normalization (TMM followed by voom)
  dge = calcNormFactors(dge)
  v = voom(dge, design, plot=TRUE)

  # Fit model to data given design
  fit = lmFit(v, design)
  fit = eBayes(fit)

  # Show top genes
  topGenes = topTable(fit, coef=ncol(design), number=100, sort.by="p")

  return(
    list(
      voomObj=v, # normalized data
      fit=fit, # fitted model and statistics
      topGenes=topGenes # the 100 most differentially expressed genes
    )
  )
}
```

```{r}
limma_pipeline2 <- function(
  tcga_data,
  condition_variable,
  reference_group = NULL,
  use_log2 = TRUE,
  min_expr = 1,
  min_samples_frac = 0.5,
  coef_index = NULL
) {
  # Step 1: Extract group information and design
  group <- factor(colData(tcga_data)[[condition_variable]])
  if (!is.null(reference_group)) {
    group <- relevel(group, ref = reference_group)
  }
  design <- model.matrix(~ group)

  # Step 2: Extract expression matrix and optionally log2-transform
  expr_matrix <- assay(tcga_data)
  if (use_log2) {
    expr_matrix <- log2(expr_matrix + 1)
  }

  # Step 3: Create DGEList (used for voom, not necessarily raw counts)
  dge <- DGEList(counts = expr_matrix,
                 samples = colData(tcga_data),
                 genes = as.data.frame(rowData(tcga_data)))

  # Step 4: Custom filtering
  keep <- rowMeans(expr_matrix > min_expr) >= min_samples_frac
  dge <- dge[keep, , keep.lib.sizes = FALSE]

  # Step 5: Normalization and voom transformation
  dge <- calcNormFactors(dge)
  v <- voom(dge, design, plot = TRUE)

  # Step 6: Linear modeling
  fit <- lmFit(v, design)
  fit <- eBayes(fit)

  # Step 7: Select coefficient for contrast
  if (is.null(coef_index)) {
    coef_index <- ncol(design)  # default to last column (typical for 2-group comparison)
  }

  topGenes <- topTable(fit,
                       coef = coef_index,
                       number = Inf,
                       sort.by = "P",
                       adjust.method = "BH")

  return(list(
    voomObj = v,
    fit = fit,
    topGenes = topGenes
  ))
}

```

```{r}
limma_res = limma_pipeline(
  tcga_data = TCGA_data,
  condition_variable="definition",
  reference_group="Solid Tissue Normal"
)
```

```{r}
saveRDS(object = limma_res,
        file = "limma_res.RDS",
        compress = FALSE)
```


```{r}
gender_limma_res = limma_pipeline(
  tcga_data=TCGA_data,
  condition_variable="gender",
  reference_group="female"
)
```
```{r}
plot_PCA = function(voomObj, condition_variable){
  group = factor(voomObj$targets[, condition_variable])
  pca = prcomp(t(voomObj$E))
  # Take PC1 and PC2 for the plot
  plot(pca$x[,1:2],col=group, pch=19)
  # include a legend for points
  legend("bottomright", inset=.01, levels(group), pch=19, col=1:length(levels(group)))
  return(pca)
}
```

```{r}
png("visuals/res_pca.png", width = 6, height = 4, units = "in", res = 300)
plot_PCA(limma_res$voomObj, "definition")  # Make sure "definition" is a valid column
dev.off()

```


```{r}
res_pca = plot_PCA(limma_res$voomObj, "gender")
```
####################################

```{r}
dataPrep_Target <- TCGAanalyze_Preprocessing(object = TCGA_data, 
                                      cor.cut = 0.6,
                                      datatype = "fpkm_unstrand",
                          width = 1000,
                          height = 1000) 
```


```{r}
library(EDASeq)
dataNorm_Target <- TCGAanalyze_Normalization(tabDF = dataPrep_Target,
                                      geneInfo = geneInfoHT,method = "geneLength")
```


```{r}
boxplot(dataPrep_Target, outline = FALSE)
boxplot(dataNorm_Target, outline = FALSE)
```

```{r}
dataFilt <- TCGAanalyze_Filtering(tabDF = dataNorm_Target,
                                  method = "quantile", 
                                  qnt.cut =  0.25)
```

```{r}
 #selection of normal samples "NT"
samplesNT <- TCGAquery_SampleTypes(barcode = colnames(dataFilt),
                                   typesample = c("NT"))
samplesNT
```


```{r}
 #selection of normal samples "NT"
samplesTP <- TCGAquery_SampleTypes(barcode = colnames(dataFilt),
                                   typesample = c("TP"))
samplesTP
```


```{r}
# Diff.expr.analysis (DEA)
dataDEGs <- TCGAanalyze_DEA(mat1 = dataFilt[,samplesNT],
                            mat2 = dataFilt[,samplesTP],
                            Cond1type = "Normal",
                            Cond2type = "Tumor",
                            fdr.cut = 0.01 ,
                            logFC.cut = 0.6,
                            method = "exactTest")
```

```{r}
fit <- limma_res$fit
deg_table <- topTable(fit, 
                      coef=ncol(fit$design), 
                      number=Inf, 
                      sort.by="p")

```


```{r}
deg_filtered <- deg_table[deg_table$adj.P.Val < 0.05 & abs(deg_table$logFC) > 0.6,]

```


```{r}
deg_table_1 <- topTable(limma_res$fit, 
                      coef=ncol(limma_res$fit$design), 
                      number=Inf, sort.by="p")

```

```{r}
deg_table_1$regulation <- "Not Significant"
deg_table_1$regulation[deg_table$logFC > 0.6 & deg_table$adj.P.Val < 0.05] <- "Upregulated"
deg_table_1$regulation[deg_table$logFC < -0.6 & deg_table$adj.P.Val < 0.05] <- "Downregulated"

# Convert to factor with ordering
deg_table_1$regulation <- factor(deg_table_1$regulation, levels = c("Upregulated", "Downregulated", "Not Significant"))

```

```{r}
library(ggplot2)

ggplot(deg_table_1, aes(x = logFC, y = -log10(adj.P.Val), color = regulation)) +
  geom_point(alpha = 0.7, size = 1.5) +
  scale_color_manual(values = c("Upregulated" = "red", 
                                "Downregulated" = "blue", 
                                "Not Significant" = "grey")) +
  geom_vline(xintercept = c(-1, 1), linetype = "dashed", color = "black") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "black") +
  theme_minimal(base_size = 14) +
  labs(title = "Volcano Plot",
       x = "Log2 Fold Change",
       y = "-Log10 Adjusted P-value",
       color = "Gene Regulation")


```

####################### Heatmap #####
```{r}
expr_matrix <- limma_res$voomObj$E
top_gene_ids <- rownames(limma_res$topGenes)
expr_top <- expr_matrix[top_gene_ids, ]
expr_scaled <- t(scale(t(expr_top)))  # Z-score per gene


```


```{r}
library(pheatmap)

# Prepare annotation
sample_annot <- as.data.frame(colData(TCGA_data)[, "definition", drop = FALSE])
colnames(sample_annot) <- "Condition"

# Save heatmap to PNG
png("visuals/heatmap_top100.png", width = 8, height = 10, units = "in", res = 600)

pheatmap(expr_scaled,
         annotation_col = sample_annot,
         show_rownames = FALSE,
         show_colnames = FALSE,
         fontsize_row = 6,
         fontsize_col = 8,
         clustering_distance_rows = "euclidean",
         clustering_distance_cols = "euclidean",
         clustering_method = "complete",
         scale = "none",  # Already scaled
         main = "Top 100 Differentially Expressed Genes",
         color = colorRampPalette(c("blue", "white", "red"))(100))

dev.off()



```


```{r}
library(clusterProfiler)
library(org.Hs.eg.db) # Human annotation
library(ReactomePA)

# Prepare gene list: only significant DEGs
deg_df <- limma_res$topGenes
deg_df <- deg_df[deg_df$adj.P.Val < 0.05 & abs(deg_df$logFC) > 1, ]

# Strip version numbers
ensembl_ids <- sub("\\..*", "", rownames(deg_df))

# Now map to Entrez IDs
entrez_ids <- bitr(ensembl_ids,
                   fromType = "ENSEMBL",
                   toType = "ENTREZID",
                   OrgDb = org.Hs.eg.db)


# KEGG enrichment
kegg <- enrichKEGG(gene = entrez_ids$ENTREZID, 
                   organism = "hsa")

# Reactome enrichment
reactome <- enrichPathway(gene = entrez_ids$ENTREZID, 
                          organism = "human")

# View results
head(kegg)
head(reactome)

# Plot
barplot(kegg, showCategory = 20)
barplot(reactome, showCategory = 20)

```


```{r}
# Top 100 upregulated
deg_up <- deg_filtered[order(deg_filtered$logFC, decreasing = TRUE), ][1:100, ]

# Top 100 downregulated
deg_down <- deg_filtered[order(deg_filtered$logFC), ][1:100, ]

# Combine
top200_deg <- rbind(deg_up, deg_down)

expr_mat <- limma_res$voomObj$E
top_expr <- expr_mat[rownames(expr_mat) %in% rownames(top200_deg), ]
top_expr <- top_expr[match(rownames(top200_deg), rownames(top_expr)), ]
expr_scaled <- t(scale(t(top_expr)))
sample_annot <- as.data.frame(colData(TCGA_data)[, "definition", drop = FALSE])
colnames(sample_annot) <- "Condition"

# Save heatmap to PNG
png("visuals/heatmap_top200.png", width = 8, height = 10, 
    units = "in", res = 600)

library(pheatmap)

top_200 <- pheatmap(expr_scaled,
         annotation_col = sample_annot,
         show_rownames = FALSE,
         show_colnames = FALSE,
         fontsize_row = 6,
         fontsize_col = 8,
         clustering_distance_rows = "euclidean",
         clustering_distance_cols = "euclidean",
         clustering_method = "complete",
         scale = "none",
         main = "Top 200 DEGs (100 Up / 100 Down)",
         color = colorRampPalette(c("blue", "white", "red"))(100))


dev.off()
print(top_200 )
```


```{r}
library(clusterProfiler)
library(org.Hs.eg.db) # Human annotation
library(ReactomePA)

# Prepare gene list: only significant DEGs
deg_df_200 <- top200_deg
deg_df_200 <- deg_df_200[deg_df_200$adj.P.Val < 0.05 & abs(deg_df_200$logFC) > 1, ]

# Strip version numbers
ensembl_ids <- sub("\\..*", "", rownames(deg_df_200))

# Now map to Entrez IDs
entrez_ids <- bitr(ensembl_ids,
                   fromType = "ENSEMBL",
                   toType = "ENTREZID",
                   OrgDb = org.Hs.eg.db)


# KEGG enrichment
kegg <- enrichKEGG(gene = entrez_ids$ENTREZID, 
                   organism = "hsa")

# Reactome enrichment
reactome <- enrichPathway(gene = entrez_ids$ENTREZID, 
                          organism = "human")

# View results
head(kegg)
head(reactome)

# Plot
barplot(kegg, showCategory = 20)
barplot(reactome, showCategory = 15)

```

#########################################################

```{r}
# To access the clinical/phenotype data (sample metadata):
clinicalTrait <- colData(TCGA_data)

# Check the structure:
#str(clinicalTrait)

# View the first few rows:
head(clinicalTrait)

```














































